### ISSUE TITLE
Docker 1.4.0 - Error loading device metadata file

### ISSUE THREAD

**Issue by phemmer** [docker/docker#9644]
After upgrading to docker 1.4.0, it refuses to start with the following error:

```
# docker -d -D                                           
INFO[0000] +job serveapi(unix:///var/run/docker.sock)   
DEBU[0000] Generated prefix: docker-254:6-67150944      
DEBU[0000] Checking for existence of the pool 'docker-254:6-67150944-pool' 
DEBU[0000] [deviceset] constructDeviceIdMap()           
DEBU[0000] Loading data for file /var/lib/docker/devicemapper/metadata/.tmp023210198 
DEBU[0000] [deviceset] constructDeviceIdMap() END       
FATA[0000] Error loading device metadata file .tmp023210198 
```

Moving that `.tmp023210198` file out of that directory allows docker to start. I'm not sure what that file is, but I'm guessing it's some sort of abandoned file docker created once upon a time.

**Comment by rhvgoyal** [comment]
@phemmer

Do you still have this .tmp023210198 file? I wanted to know what are the contents of the file.

During docker start we construct a device map from all the existing metadata files. So looks like somehow this rouge file showed up in that directory and could not be parsed.

Either there is a problem with logic which constructs device Id map or there is a bug somewhere else which drops this tmp file in metadata dir.

Can you reproduce it?

**Comment by phemmer** [comment]
@rhvgoyal the file was empty.

As far as reproduction, no. I stopped using the devicemapper driver a couple months ago because I was constantly running into issues with it.

**Comment by rhvgoyal** [comment]
Following patch should fix it. I am testing it and will send a pull request this afternoon.

devicemapper: Skip the files with prefix "." during device map construction

Any file which starts with "." is not a valid metadata file. Skip it
during device map construction.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>
---
 daemon/graphdriver/devmapper/deviceset.go |    5 +++++
 1 file changed, 5 insertions(+)

Index: docker/daemon/graphdriver/devmapper/deviceset.go
===================================================================
--- docker.orig/daemon/graphdriver/devmapper/deviceset.go       2015-01-22 09:22:23.309181174 -0500
+++ docker/daemon/graphdriver/devmapper/deviceset.go    2015-01-29 12:02:26.482687104 -0500
@@ -323,6 +323,11 @@ func (devices *DeviceSet) deviceFileWalk
                return nil
        }
+       if strings.HasPrefix(finfo.Name(), ".") {
+               log.Debugf("Skipping file %s", path)
+               return nil
+       }
+
        if finfo.Name() == deviceSetMetaFile {
                log.Debugf("Skipping file %s", path)
                return nil

**Comment by rhvgoyal** [comment]
Hmm.., looks like this interface is not friendly to pasting patches. Anyway, I will send a pull request this afternoon.

**Comment by stevenschlansker** [comment]
Note that in my case, the erroring file did not begin with `.` so I do not think this patch fixes all possible problems.

**Comment by phemmer** [comment]
Is there any reason not being able to load a file in this directory is a fatal error? Can it not just complain, and keep going?

**Comment by rhvgoyal** [comment]
Generated a pull request for this issue.

https://github.com/docker/docker/pull/10440

**Comment by rhvgoyal** [comment]
This temp file might have been created by docker itself as it created temp file in metadata dir. It is possible that docker created a temp file and then died. In that case above pull request should take care of all these temp files which start with "."

// Given json data and file path, write it to disk
func (devices *DeviceSet) writeMetaFile(jsonData []byte, filePath string) error {
        tmpFile, err := ioutil.TempFile(devices.metadataDir(), ".tmp")
        if err != nil {
                return fmt.Errorf("Error creating metadata file: %s", err)
        }

**Comment by akram** [comment]
Hi 

Pull Request 10440 was supposed to fix the issue? wasn't it ?
I am still facing it with docker 1.5

```
[root@172 tools]# docker version
Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.3.3
Git commit (client): a8a31ef/1.5.0
OS/Arch (client): linux/amd64
FATA[0000] Get http:///var/run/docker.sock/v1.17/version: read unix /var/run/docker.sock: connection reset by peer. Are you trying to connect to a TLS-enabled daemon without TLS?
[root@172 tools]# docker ps
FATA[0000] Cannot connect to the Docker daemon. Is 'docker -d' running on this host?
[root@172 tools]# journalctl -u docker -f
-- Logs begin at Fri 2015-02-27 01:10:40 UTC. --
Feb 27 10:33:20 172.16.50.224 systemd[1]: Starting Docker Application Container Engine...
Feb 27 10:33:20 172.16.50.224 docker[17464]: time="2015-02-27T10:33:20Z" level="info" msg="+job serveapi(unix:///var/run/docker.sock)"
Feb 27 10:33:20 172.16.50.224 docker[17464]: time="2015-02-27T10:33:20Z" level="info" msg="Listening for HTTP on unix (/var/run/docker.sock)"
Feb 27 10:33:20 172.16.50.224 docker[17464]: time="2015-02-27T10:33:20Z" level="fatal" msg="Error loading device metadata file .tmp650769236"
Feb 27 10:33:20 172.16.50.224 systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Feb 27 10:33:20 172.16.50.224 systemd[1]: Failed to start Docker Application Container Engine.
Feb 27 10:33:20 172.16.50.224 systemd[1]: Unit docker.service entered failed state.
Feb 27 10:33:20 172.16.50.224 systemd[1]: Starting Docker Application Container Engine...
Feb 27 10:33:20 172.16.50.224 systemd[1]: docker.service start request repeated too quickly, refusing to start.
Feb 27 10:33:20 172.16.50.224 systemd[1]: Failed to start Docker Application Container Engine.
```

I still have the tmp file present, if it helps for debugging let me know

**Comment by rhvgoyal** [comment]
I just tested latest docker and it seems to work fine for me. For testing I manually created a file call .tmp1234 in /var/lib/docker/devicemapper/metadata dir. And see following message in logs when docker starts.

Feb 27 09:54:16 tiger docker[7711]: time="2015-02-27T09:54:16-05:00" level="debug" msg="Skipping file /var/lib/docker/devicemapper/metadata/.tmp1234"

Are you sure that docker instance you are running contains my fix.

**Comment by rhvgoyal** [comment]
@akram 

Looks like 1.5.0 does not have my commit. It went in after that. That explains why you are still facing this issue.

**Comment by Code-Vortex** [comment]
for me solved by removing the file which was unable to be loaded

this was in the result of service docker status

```
Feb 28 16:11:56 localhost.localdomain docker[16862]: time="2015-02-28T16:11:56+02:00" level="fatal" msg="Error loading device metadata file ....859135"
```

solved by the following:

```
sudo find /var/lib/docker/ -name "*859135*"
sudo rm /var/lib/docker/devicemapper/metadata/.tmp361859135
```

**Comment by thaJeztah** [comment]
@rhvgoyal should this be marked "resolved"?  If I understand correctly, it's resolved in master and will be fixed in the next release (either 1.5.1 or 1.6.0, whatever comes first)?

**Comment by rhvgoyal** [comment]
@thaJeztah 

Yes, this is resolved upstream now. Following is the commit. I think fix should show up in 1.5.1.

commit 080a6f1e4b7c33ce9b7b1b90cbf8fd90658226c2
Author: Vivek Goyal <vgoyal@redhat.com>
Date:   Thu Jan 29 13:19:18 2015 -0500

    devicemapper: Skip the files with prefix "." during device map construction
    
    Any file which starts with "." is not a valid metadata file. Skip it
    during device map construction.
    
    Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

**Comment by thaJeztah** [comment]
Thanks @rhvgoyal!

@phemmer can you close the issue? Thanks!

**Comment by camjackson** [comment]
I've been getting this error on docker 1.5, but the metadata file being referred to was not a .tmp one. They were ones like `/var/lib/docker/devicemapper/metadata/1f374ad620865cec9af566778d8834ad7f1d7ea6ffbe896ff289014ae2ca479a`.

The ones being complained about seem to be corrupt, as `less` warned me that they were binary, and looking at them they were just filled with garbage. All I did was remove them from that directory (a few times, as docker complained about 5 of them), and it eventually worked. Just posting this in case it helps someone else.

**Comment by thaJeztah** [comment]
@camjackson given the discussion above (https://github.com/docker/docker/issues/9644#issuecomment-76725711), the issue discussed here will be fixed in the upcoming 1.6.0 release.

You can try this by downloading the current development version from https://master.dockerproject.com. Note that the development version may contain bugs, so *don't* run it in a production environment.

**Comment by rhvgoyal** [comment]
@camjackson 

File name seems fine and devicemapper graphdriver will try to parse it during startup. If somehow file got corrupted and can't be parsed, then docker will fail to start.

So real problem here is to figure out how did these files get corrupted in the first place. Can you reproduce it?

**Comment by camjackson** [comment]
@rhvgoyal Yeah, the files were definitely corrupt, no idea how they got like that though. It seemed to be the upgrade from 1.4 to 1.5 that did it, but I can't reproduce it now.

**Comment by pborisov-rmn** [comment]
@rhvgoyal , @phemmer : I've been able to reproduce docker being unable to load a device metadata file after having the host machine had its disk fill. I was running docker 1.5.0. Sadly I didn't keep the file itself, but I was able to get docker to start up again by following the procedure @sh3llsh0ck outlined.

I'll be dealing with the issue by avoiding filling up the disk again and upgrading at the next opportunity.
